File: Lab1/C/src/qsort.c
#include<stdio.h>
#include<time.h>

int a[5000005] = {0};

void qsort(int l,int r){
    if(l>=r){
        return;
    }
    int i = l, j = r, x = a[l];
    while(i<j){
        while(i<j && x <= a[j]){
            j--;
        }
        a[i] = a[j];
        while(i<j && a[i] < x){
            i++;
        }
        a[j] = a[i];
    }
    a[j]= x;
    qsort(l,j-1);
    qsort(j+1,r);
    return;
}


int main(){
    int num;
    // Read data file
    FILE* fp=NULL;
    if((fp = fopen("../../5000000.txt","r"))==NULL){
        printf("File not exist\n");
        return 1;
    }
    fscanf(fp,"%d",&num);
    for(int i=0;i<num;i++){
        fscanf(fp,"%d",&a[i]);
    }
    fclose(fp);

    // Print Before sort
    printf("Before sort(top 10): ");
    for(int i=0;i<10;i++){
        printf("%d ",a[i]);
    }
    printf("\n\n");

    // Sort
    clock_t st,et;
    printf("Start\n");
    st = clock();
    qsort(0,num-1);
    et = clock();
    printf("Finish!\n\n");

    
    // Print After sort
    printf("After sort(top 10): ");
    for(int i=0;i<10;i++){
        printf("%d ",a[i]);
    }
    printf("\n\n");
    printf("run time: %f\n",(double)(et-st)/(CLOCKS_PER_SEC));
}

File: Lab1/Haskell/src/qsort.hs
import System.IO

into:: String->Int
into = read

qsort :: (Ord a)=>[a]->[a]
qsort [] =[]
qsort (x:xs)=
    let smaller=[a|a<-xs,a<=x]
        bigger=[a|a<-xs,a>x]
    in qsort smaller ++[x]++qsort bigger

main = do
    --  Read data file
    fp <-openFile "5000000.txt" ReadMode
    cont <- hGetContents fp
    let text = lines cont
    let t2 = text!!1
    let a = map read $words t2::[Int]

    -- Print Before sort
    putStrLn("Before sort(top 3): ")
    print(a!!0)
    print(a!!1)
    print(a!!2)
    putStrLn("")

    -- Sort
    putStrLn("Start")
    let b = qsort(a)
    putStrLn("Finish!")
    putStrLn("")

    -- Print After sort
    putStrLn("After sort(top 3): ")
    print(b!!0)
    print(b!!1)
    print(b!!2)